File: rust/src/day10.rs
use std::collections::HashSet;

use array2d::Array2D;

use crate::util::{self, grid_to_a2d, Dir, Pos, EAST, NORTH, SOUTH, WEST};

pub const NAME: &str = "Day 10: Pipe Maze";

pub fn _example() -> String {
    String::from("")
}

pub fn data() -> String {
    util::get_input("day10.txt")
}

#[cfg(test)]
mod test_result {
    use super::*;
    #[test]
    fn part01() {
        assert_eq!(super::part01(data()), 0);
    }
    #[test]
    fn part02() {
        assert_eq!(super::part02(data()), 0);
    }
}
#[cfg(test)]
mod test_example {
    use super::*;
    #[test]
    fn part01() {
        assert_eq!(super::part01(_example()), 0);
    }
    #[test]
    fn part02() {
        assert_eq!(super::part02(_example()), 0);
    }
}

fn next_dir(d: Dir, t: (Dir, Dir)) -> Option<Dir> {
    if t.0 == -d {
        return Some(t.1);
    }
    if t.1 == -d {
        return Some(t.0);
    }
    None
}

fn bends(a: &Array2D<char>, p: Pos) -> (Dir, Dir) {
    match a[p.as_a2d_index()] {
        '|' => (NORTH, SOUTH),
        '-' => (EAST, WEST),
        'L' => (NORTH, EAST),
        'J' => (NORTH, WEST),
        '7' => (SOUTH, WEST),
        'F' => (SOUTH, EAST),
        _ => panic!("p"),
    }
}

fn flips(c: char) -> bool {
    match c {
        '|' => true,
        '-' => false,
        'L' => false,
        'J' => false,
        '7' => true,
        'F' => true,
        _ => panic!("p"),
    }
}

fn find_start(a: &Array2D<char>) -> Pos {
    a.enumerate_row_major()
        .filter_map(|(p, c)| match *c == 'S' {
            true => Some(Pos::from_a2d_index(p)),
            false => None,
        })
        .next()
        .unwrap()
}

fn find_start_dir(a: &Array2D<char>, start_pos: Pos) -> Dir {
    [NORTH, SOUTH, EAST, WEST]
        .iter()
        .filter_map(|d| next_dir(*d, bends(&a, start_pos + *d)))
        .next()
        .unwrap()
}
pub fn part01(data: String) -> usize {
    let a = grid_to_a2d(&data);
    let start_pos = find_start(&a);
    let mut dir = find_start_dir(&a, start_pos);
    let mut pos = start_pos;
    let mut n = 0;
    loop {
        pos = pos + dir;
        n += 1;
        if pos == start_pos {
            break;
        };
        dir = next_dir(dir, bends(&a, pos)).unwrap();
    }
    n / 2
}

pub fn part02(data: String) -> usize {
    let mut a = grid_to_a2d(&data);
    let mut visited = HashSet::with_capacity(10000);
    let start_pos = find_start(&a);
    let mut dir = find_start_dir(&a, start_pos);
    a.set(
        start_pos.0 as usize,
        start_pos.1 as usize,
        match next_dir(SOUTH, bends(&a, start_pos + SOUTH)) {
            Some(_) => '|',
            None => '-',
        },
    )
    .unwrap();
    let mut pos = start_pos;
    loop {
        visited.insert(pos.as_a2d_index());
        pos = pos + dir;
        if pos == start_pos {
            break;
        };
        dir = next_dir(dir, bends(&a, pos)).unwrap();
    }
    let mut n = 0;
    let mut is_inside: bool;
    for (row_num, row) in a.rows_iter().enumerate() {
        is_inside = false;
        for (col_num, c) in row.enumerate() {
            match visited.get(&(row_num, col_num)) {
                Some(_) => {
                    if flips(*c) {
                        is_inside = !is_inside
                    }
                }
                None => {
                    if is_inside {
                        n += 1
                    }
                }
            }
        }
    }
    n
}

File: haskell/src/Day10.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE QuasiQuotes #-}
{-# OPTIONS_GHC -Wno-incomplete-patterns -Wunused-top-binds #-}
{-# OPTIONS_GHC -Wno-unused-top-binds #-}

module Day10 (runme, runex) where

import Text.RawString.QQ

import Data.ByteString (ByteString)
import qualified Data.ByteString.Char8 as B
import Data.Maybe (mapMaybe, fromJust)

import RunUtil (RunMe, runMeByteString)
import AOCHelper (readInpByteSTring)
import qualified BSArray as BS
import qualified Data.HashSet as S
import Debug.Trace (trace)
import Data.List (foldl')

type MySet = S.HashSet (Int,Int)

example :: ByteString
example =
  [r|7-F7-
.FJ|7
SJLL7
|F--J
LJ.LJ
|]

runex :: RunMe
runex =
  runMeByteString
    "Day 10 - example"
    (return example)
    part1
    (Just 8)
    part2
    (Just 1)

runme :: RunMe
runme =
  runMeByteString
    "Day 10: Pipe Maze (TODO: try intset)"
    (readInpByteSTring "day10.txt")
    part1
    (Just 7005)
    part2
    (Just 417)

---
newtype Pos = Pos (Int, Int) deriving (Eq, Show)
data Dir = NORTH | EAST | SOUTH | WEST deriving (Eq, Show)

(.+.) :: Pos -> Pos -> Pos
(.+.) (Pos (a,b)) (Pos (c,d)) = Pos (a+c, b+d)

toPos :: (Int, Int) -> Pos
toPos (a,b) = Pos (a,b)

fromPos (Pos (a,b)) = (a, b)

dToPos :: Dir -> Pos
dToPos NORTH = Pos (-1,0)
dToPos SOUTH = Pos (1, 0)
dToPos EAST = Pos (0,1)
dToPos WEST = Pos (0,-1)

(.->.) :: Pos -> Dir -> Pos
(.->.) p d = p .+. dToPos d
-- 


turn :: Dir -> Char -> Maybe Dir
turn d c = case (d, c) of
  (NORTH, '|') -> Just NORTH
  (SOUTH, '|') -> Just SOUTH
  (WEST,  '-') -> Just WEST
  (EAST,  '-') -> Just EAST
  (EAST,  '7') -> Just SOUTH
  (NORTH, '7') -> Just WEST
  (NORTH, 'F') -> Just EAST
  (WEST,  'F') -> Just SOUTH
  (SOUTH, 'L') -> Just EAST
  (WEST,  'L') -> Just NORTH
  (EAST,  'J') -> Just NORTH
  (SOUTH, 'J') -> Just WEST
  _ -> Nothing


turnandmove :: Char -> (Pos, Dir) -> Maybe (Pos, Dir)
turnandmove 'S' (p, d) = Just  (p .->. d , d)
turnandmove c (p,d) =  (\d' -> (p .->. d' , d')) <$> turn d c

loop :: BS.BSArray -> [Pos] -> Pos -> Dir -> Maybe [Pos]
loop bs l p d
      | c == 'S' && l /= [] = Just l
      | otherwise = case turnandmove c (p,d) of
      Nothing -> Nothing
      Just (p',d') ->  loop bs (p:l) p' d'
  where c = BS.lookup bs . fromPos $ p

looplist :: BS.BSArray -> Pos -> [Pos]
looplist bs start = head . mapMaybe (loop bs [] start) $ [NORTH, EAST, SOUTH, WEST]

part1 :: ByteString -> IO Integer
part1 s = do
  let bs = BS.makeBSarray s
  let start = toPos . fromJust . BS.elemIndex bs $ 'S'
  return . toInteger
         . flip div 2
         . length
         $ looplist bs start

countrow :: Bool -> MySet -> Int -> ByteString -> Int
countrow sflips m rownum s = fst . foldl' f (0, False) $ zip [0..] (B.unpack s)
  where f (count, isinside) (colnum, c) =
          if (rownum, colnum) `S.member` m
            then (count, if flipit then not isinside else isinside)
            else (if isinside then count+1 else count, isinside)
          where flipit = c == '|'
                      || c == '7'
                      || (c == 'S' && sflips)
                      || c == 'F'

part2 :: ByteString -> IO Integer
part2 s = do
    let bs = BS.makeBSarray s
    let start = toPos . fromJust . BS.elemIndex bs $ 'S'
    let looppoints = S.fromList . map fromPos $ looplist bs start
    let sflips = fromPos (start .->. SOUTH) `S.member` looppoints

    return . toInteger
           . sum
           . map (\rownum -> countrow sflips looppoints rownum (BS.row bs rownum))
           $  [0..(BS.rows bs -1)]
